# Original Code

**The following code will compute default MICE over the mtcars_missing dataset:**

In [ ]:
# Load libraries
import numpy as np
import pandas as pd
import io




# Read file
from google.colab import files
uploaded = files.upload()
dat = pd.read_csv(io.BytesIO(uploaded['mtcars_missing.csv']), sep = ",")


# Select numerical variables
numerical_vars = list(set(dat.columns) - set(['model']))


# Import MICE method
!pip install fancyimpute
from fancyimpute import IterativeImputer as MICE


# 3) Define "model"
model = MICE()


# 4) Train "model"
model.fit(dat[numerical_vars])


# 5) "Predict"
dat[numerical_vars] = model.transform(dat[numerical_vars])

Saving mtcars_missing.csv to mtcars_missing.csv
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.7/154.7 kB 4.6 MB/s eta 0:00:00
  Created wheel for fancyimpute: filename=fancyimpute-0.7.0-py3-none-any.whl size=29880 sha256=6fdcec08e3eaa2e0f2b3f17b190673f38cf63eaaba84a52b15f8c169eb82cffd
  Stored in directory: /root/.cache/pip/wheels/7b/0c/d3/ee82d1fbdcc0858d96434af108608d01703505d453720c84ed
  Created wheel for knnimpute: filename=knnimpute-0.1.0-py3-none-any.whl size=11329 sha256=6b8a875be904a81f2334e95f73e78983e92909a8de9fdaf7b8d4667130c38220
  Stored in directory: /root/.cache/pip/wheels/46/06/a5/45a724630562413c374e29c08732411d496092408b3a7bf754
Successfully built fancyimpute knnimpute


Let's see the results

This is the original dataset.

In [ ]:
old_dat = pd.read_csv(io.BytesIO(uploaded['mtcars_missing.csv']), sep = ",")
old_dat.head()

model   mpg  cyl   disp   hp  drat     wt   qsec  vs   am  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0  NaN   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0  1.0   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1  1.0   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1  0.0   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0  0.0   

   gear  carb  
0     4     4  
1     4     4  
2     4     1  
3     3     1  
4     3     2

This is the dataset after running the code and applying MICE.

In [ ]:
dat.head()

model   mpg  cyl   disp     hp  drat     wt   qsec   vs  \
0          Mazda RX4  21.0  6.0  160.0  110.0  3.90  2.620  16.46  0.0   
1      Mazda RX4 Wag  21.0  6.0  160.0  110.0  3.90  2.875  17.02  0.0   
2         Datsun 710  22.8  4.0  108.0   93.0  3.85  2.320  18.61  1.0   
3     Hornet 4 Drive  21.4  6.0  258.0  110.0  3.08  3.215  19.44  1.0   
4  Hornet Sportabout  18.7  8.0  360.0  175.0  3.15  3.440  17.02  0.0   

         am  gear  carb  
0  0.610321   4.0   4.0  
1  1.000000   4.0   4.0  
2  1.000000   4.0   1.0  
3  0.000000   3.0   1.0  
4  0.000000   3.0   2.0

We can see that in the first row, the original missing value of *am* has been imputed as 0.610321.

# Exercise

**Modify the code to make MICE apply a Decision Tree model to impute missing values (you should modify line model = MICE() to select the model you want to apply in MICE).**

In [ ]:
# Read file
dat = pd.read_csv(io.BytesIO(uploaded['mtcars_missing.csv']), sep = ",")


# Select numerical variables
numerical_vars = list(set(dat.columns) - set(['model']))


# Import MICE method
from fancyimpute import IterativeImputer as MICE


# 3) Define "model"
from sklearn.tree import DecisionTreeRegressor # ADDED!!!
model = MICE(DecisionTreeRegressor()) # MODIFIED!!!


# 4) Train "model"
model.fit(dat[numerical_vars])


# 5) "Predict"
dat[numerical_vars] = model.transform(dat[numerical_vars])

**Check the differences in the imputed values. Which one looks a better approach?**

Let's analyze the new dataset.

In [ ]:
dat.head()

model   mpg  cyl   disp     hp  drat     wt   qsec   vs   am  \
0          Mazda RX4  21.0  6.0  160.0  110.0  3.90  2.620  16.46  0.0  1.0   
1      Mazda RX4 Wag  21.0  6.0  160.0  110.0  3.90  2.875  17.02  0.0  1.0   
2         Datsun 710  22.8  4.0  108.0   93.0  3.85  2.320  18.61  1.0  1.0   
3     Hornet 4 Drive  21.4  6.0  258.0  110.0  3.08  3.215  19.44  1.0  0.0   
4  Hornet Sportabout  18.7  8.0  360.0  175.0  3.15  3.440  17.02  0.0  0.0   

   gear  carb  
0   4.0   4.0  
1   4.0   4.0  
2   4.0   1.0  
3   3.0   1.0  
4   3.0   2.0

We can see that now in the first row the original missing value of *am* has been imputed as 1.0. This may seem more reasonable, as *am* is a binary dummy variable.